### PR TITLE
Start using cookies instead of localStorage by default

### LINKED ISSUE #402: localStorage is spammed with state/nonce combos
Every time before `auth0.js` performs a redirect to the `/authorize` endpoint, `state` and `nonce` are generated and stored in `localStorage` for verification upon return. If the user lands on the Auth0 hosted login page but does not login, the stored `state/nonce` combo won't be removed.

This means that `localStorage` can fill up with lots of unused values, especially during development.
```
com.auth0.auth.3z-j16G6OFMF_v-nt_QUZS5VV~vf.fGD:	{"nonce":"55yFu~kZUv0KT6azRtCIyf_c3dx-GAmJ"}
```
image

Is there a reason why this could not be done via a single `localStorage` key, with JSON data?
Or possibly that in addition to `nonce`, a `issuedAt` field could be used for cleaning up old `state/nonce` combos?

### PR DESCRIPTION
**Pull request by luisrudge**
fix #402 #816 

In some cases, there's no way to old local storage entries because we don't have enough information:

- if the auth result wasn't successful
- if the cross origin auth was successful, but it didn't use the cross auth callback page

Because of that, cookies are a better fit for storage, since it has built-in expiration dates.

There's a new option `__tryLocalStorageFirst` that restores the previous behavior if you find any compatibility issues (please report them if you find it) or prefer to use localStorage for some reason.

This is not considered a breaking change because there shouldn't be any behavioral changes between versions.

### CODE REVIEW

File: src/helper/storage/handler.js [hunk 1]
```diff
@@ -3,9 +3,12 @@ import DummyStorage from './dummy';
 import CookieStorage from './cookie';
 import Warn from '../warn';
 
-function StorageHandler() {
+function StorageHandler(options) {
   this.warn = new Warn({});
   this.storage = new CookieStorage();
+  if (!options.__tryLocalStorageFirst) {
```
Comment (joshcanhelp) [inline src/helper/storage/handler.js]: Do we want to be explicit about what `options.__tryLocalStorageFirst` can be? This is truth-y, should it be `if (!options.__tryLocalStorageFirst === true)` instead? Edge case and a new option so no existing behavior but `"no"` or `"false"` would trigger a positive case here. Just wondering how this check is done elsewhere.

File: test/helper/storage-handler.test.js [hunk 2]
```diff
@@ -55,30 +59,30 @@ describe('helpers storage handler', function() {
...
       var key = 'foo';
-      var handler = new StorageHandler();
+      var handler = new StorageHandler({ __tryLocalStorageFirst: true });
       handler.removeItem(key);
       expect(removeItemSpy.firstCall.args).to.be.eql([key]);
     });
   });
   describe('should use cookie storage', function() {
-    it('when localstorage is not available', function() {
+    it('when __tryLocalStorageFirst is true but localSTorage is not available', function() {
```
Comment (joshcanhelp) [inline test/helper/storage-handler.test.js]: "localStorage"

File: src/helper/ssodata.js [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
-    if (!ssodata) {
-      return;
-    }
-    return JSON.parse(ssodata);
+function SSODataStorage(options) {
+  this.storage = new Storage(options);
+}
+
+SSODataStorage.prototype.set = function(connection, sub) {
+  var ssodata = {
```
[the remark this answers is not in the document]
Comment (luisrudge) [reply src/helper/ssodata.js]: auth0.js is not transpiling code yet, so no let or const are allowed.

File: test/helper/storage-handler.test.js (hunk 2 above)
Comment (luisrudge) [reply test/helper/storage-handler.test.js]: CAN I USE ALL CAPS LIKE THIS LOCALSTORAGE WHAT DO YOU THINK

File: src/helper/storage/handler.js (hunk 1 above)
Comment (luisrudge) [reply src/helper/storage/handler.js]: done

File: test/helper/storage-handler.test.js (hunk 2 above)
Comment (luisrudge) [reply test/helper/storage-handler.test.js]: fixed 😬

Comment (joshcanhelp) [reply test/helper/storage-handler.test.js]: Either all caps or camel, pick one! 😆

**Comment by mythmon** [comment]
> This is not considered a breaking change because there shouldn't be any behavioral changes between versions.

I have a question about this. Before this change, auth details were stored in localstorage, and kept entirely on the client. After this change, the auth details are stored in cookies. If I understand correctly, all of the cookies on the client are sent to the server for each request. This bloats the request size, and could represent a credentials leak.

Is my understanding above wrong? If not, this does not seem a change without behavioral consequences. I was extremely surprised to see it in a minor release of the library.

**Comment by luisrudge** [comment]
There are no auth details stored in cookies. We only store:

One cookie for (expires in 24 hours):
- last used connection
- last user id that logged in

One cookie for (expires in 30 minutes):
- nonce (auto generated or provided by the developer)
- state (auto generated or provided by the developer)

One cookie for (expires in 15 minutes):
- cross origin verification key

Yes, the cookies increase the overall request size and that's why we made the transactional cookies short lived. We considered that slightly increasing the payload size was a better trade off than possibly making our customers' apps crash due to localStorage being full.

**Comment by sholladay** [comment]
FYI, cookies are _more_ prone to size and storage quantity limitations.
 - http://browsercookielimits.squawky.net/
 - https://stackoverflow.com/questions/2989284/what-is-the-max-size-of-localstorage-values

One notable difference is that if cookie storage is full, it won't cause any JavaScript errors (to my knowledge), whereas `locatStorage.setItem()` will throw. I would probably prefer the `localStorage` exception with a known error type (which can be ignored explicitly) over a silent failure. Maybe there are other good reasons to use cookies, though, like the expiration system.

File: src/helper/storage/handler.js [hunk 4]
```diff
@@ -3,9 +3,12 @@ import DummyStorage from './dummy';
 import CookieStorage from './cookie';
 import Warn from '../warn';
 
-function StorageHandler() {
+function StorageHandler(options) {
   this.warn = new Warn({});
   this.storage = new CookieStorage();
+  if (options.__tryLocalStorageFirst !== true) {
```
Comment (jbeurel) [inline src/helper/storage/handler.js]: Hello @luisrudge. How can I use LocalStorage instead of CookieStorage in a TypeScript application? It seems impossible to set __tryLocalStorageFirst to `true` because of `AuthorizeOptions` or `AuthOptions` type definition: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/auth0-js/index.d.ts#L166

Comment (luisrudge) [reply src/helper/storage/handler.js]: send a PR to the typing repo adding this flag?

Comment (luisrudge) [reply src/helper/storage/handler.js]: Also, why you want to use localStorage instead of cookies? that would be pretty helpful to understand, since we're considering removing localStorage in the future.

Comment (jbeurel) [reply src/helper/storage/handler.js]: Hello @luisrudge

"send a PR to the typing repo adding this flag?", Yes. PR done and merged by DefinitelyTyped team 👍

"Also, why you want to use localStorage instead of cookies?" Because I did not find how to make auth0 cookie work with Electron.

**Comment by luuuis** [comment]
One customer was concerned about these cookies.

> Customer [...] has noticed a cookie appearing on their domain, com.auth0.auth.testNonce3 - what is this used for? Is it a test that has leaked onto production somehow?

https://auth0team.atlassian.net/projects/ESD/queues/custom/405/ESD-1523
